Initial draft of Design for extensions (#60)
opensearch-project/opensearch-sdk: DESIGN.md

    port: '4532' // port to reach
    version: '1.0' // extension version
    description: Extension for the Opensearch SDK Repo // description of the extension
    opensearchVersion: '3.0.0' // OpenSearch compatibility
```

### Communication
As we are running extensions on the port defined in the `extensions.yml`, the communication between OpenSearch and Extensions happens using a ServerSocket which binds the port and the host address. OpenSearch will initialize the extensions during the bootstrap by making a request to all the extensions running on different ports and thus creating a medium for the future requests.

### OpenSearch SDK
Currently, plugins relies on extension points to communicate with OpenSearch. To turn plugins into extensions, all the extension points should be converted into Transport APIs which will be present in the SDK. Plugins need to integrate SDK, call those APIs, and later SDK will take care of the communication and the required attributes from OpenSearch.

### Settings
Walking through a similar example as plugin above, after extension registration is done, extension makes an API call to register custom settings to OpenSearch.

> `ExtensionsOrchestrator` receives the requests, forwards it to `SettingsModule` to register a new setting and wala, the user is now able to toggle the setting via `_settings` Rest API.

Reviewer: s/wala/voilà/

    uniqueId: opensearch-sdk-1 // identifier for the extension
    hostName: 'sdk_host' // name of the host where extension is running
    hostAddress: '127.0.0.1' // host to reach
    port: '4532' // port to reach
    version: '1.0' // extension version
    description: Extension for the Opensearch SDK Repo // description of the extension
    opensearchVersion: '3.0.0' // OpenSearch compatibility	
```

### Communication
As we are running extensions on the port defined in the `extensions.yml`, the communication between OpenSearch and Extensions happens using a ServerSocket which binds the port and the host address. OpenSearch will initialize the extensions during the bootstrap by making a request to all the extensions running on different ports and thus creating a medium for the future requests.

### OpenSearch SDK
Currently, plugins relies on extension points to communicate with OpenSearch. To turn plugins into extensions, all the extension points should be converted into Transport APIs which will be present in the SDK. Plugins need to integrate SDK, call those APIs, and later SDK will take care of the communication and the required attributes from OpenSearch.

> ### Latency

Author: @owaiskazi19 lets at Latency as an FAQ.

# Extensions

*Note*: This document is evolving and is in draft state.

Plugin architecture enables extending core features of OpenSearch. There are various kinds of plugins which are supported.
But, the architecture has significant problems for OpenSearch customers. Importantly, plugins can fatally impact the cluster
i.e critical workloads like ingestion/search traffic would be impacted because of a non-critical plugin like s3-repository failed with an exception.

This problem is exponentially grows when we would like to run a 3rd Party plugin from the community.  
As OpenSearch and plugins run in the same process, it brings in security risk, dependency conflicts and reduces the velocity of releases.

> Introducing extensions, a simple and easy way to extend features of OpenSearch. It would support all plugin features and enable them to run in a different process or another node via OpenSearch SDK.

Reviewer 2: ```suggestion
Introducing extensions, a simple and easy way to extend features of OpenSearch. It would support all plugin features and enable them to run in a separate process or on another node via OpenSearch SDK.
```

Meta Issue: [Steps to make OpenSearch extensible](https://github.com/opensearch-project/OpenSearch/issues/2447)  
Sandboxing: [Step towards modular architecture in OpenSearch](https://github.com/opensearch-project/OpenSearch/issues/1422)  
Security: [Security for extensions](SECURITY.md) 

## Plugins Architecture

![](Docs/plugins.png)

> Plugins are installed via [`opensearch-plugin`](https://github.com/opensearch-project/OpenSearch/blob/main/distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java) and are class loaded into OpenSearch.

Reviewer 2: ```suggestion
Plugins are installed via [`opensearch-plugin`](https://github.com/opensearch-project/OpenSearch/blob/main/distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java) plugin and are class loaded into OpenSearch.
```
We can link any plugin's build.gradle file here for more visibility having opensearch-plugin

Author: This is not the gradle plugin. This is `opensearch-plugin` binary tool which is already linked here. 
Are you looking for something else?

> Plugins run within OpenSearch as a single process. Plugins interface with OpenSearch via extension points which plug into the core modules of OpenSearch.

Reviewer 3: Would love to see more depth around `extension point`, is there another doc to link to or to be authored?

Reviewer 2: More about extension points is present [here](https://opensearch.org/blog/technical-post/2021/12/plugins-intro/)

Author: This blog post should cover all the information, and its also linked in the design doc. 
Let me know if you think we need more details.

This [blog post](https://opensearch.org/blog/technical-post/2021/12/plugins-intro/) helps untangle how plugins work. 

Walking through an example, a Plugin would like to register a custom setting which could be toggled via Rest API by the user. 
The plugin uses compiles with OpenSearch x.y.z version and generates a `.zip`.   
This `.zip` file is installed via `opensearch-plugin` tool which unpacks the code and places it under `~/plugins/<plugin-name>`.
During the bootstrap of OpenSearch node, it class loads all the code under `~/plugins/` directory. `Node.java` makes a call to get all settings the plugins would like to register. These settings are used as `additionalSettings` and construct `SettingsModule` instance which tracks all settings.

## Extensions Architecture

> ![](Docs/Extensions.png)

Reviewer 2: Thanks for the design architecture. Looks really good :)

Reviewer 2: Do you think we should add SDK in Extension box in the diagram? Just to give a better clarity on how SDK is making the plugin turn as an extension

Author: Haha sure. I actually did but somehow I linked an earlier version. Will update it.

Extensions are independent processes which are built using `opensearch-sdk`. They communicate with OpenSearch via [transport](https://github.com/opensearch-project/OpenSearch/tree/main/modules/transport-netty4) protocol which today is used to communicate between OpenSearch nodes.  

> Extensions are designed to extend features via APIs which are exposed using extension points of OpenSearch.  

Reviewer 2: ```suggestion
Extensions are designed to extend features via Transport APIs which are exposed using extension points of OpenSearch.  
```

### Discovery

> Extensions are discovered and configured via `extensions.yml` which is read by OpenSearch during the node bootstrap. `ExtensionsOrchestrator` reads through the config file at `~/extensions` and registers extensions within OpenSearch.

Reviewer 2: ```suggestion
Extensions are discovered and configured via `extensions.yml`, same way we currently have `plugin-descriptor.properties` which is read by OpenSearch during the node bootstrap. `ExtensionsOrchestrator` reads through the config file at `~/extensions` and registers extensions within OpenSearch.
```
Feel free to change it

Here is an example extension configuration `extensions.yml`:

```

> extensions:

Reviewer 3: I would find it very useful to have inline comments explain the usage behind these settings.

Reviewer 3: This still hasn't been updated

Author: Added comments. Let me know if that helps.

Author: Now it should be :) for real.

  - name: opensearch-sdk
    uniqueId: opensearch-sdk-1
    hostName: 'sdk_host'
    hostAddress: '127.0.0.1'
    port: '4532'
    version: '1.0'
    description: Extension for the Opensearch SDK Repo

>     opensearchVersion: '3.0.0'

Reviewer 3: Why does the extension need to provide an opensearch version in this configuration?  Shouldn't this be part of the handshake during the discovery/initialization process?

Reviewer 4: I think this is because in future a given extension can work with several versions of OpenSearch and not tied to one, this is around API versioning. Hence providing input on what version of OpenSearch a given extension can work would need to be specified.

Author: Thats a good point. For now its static, and configured via `yml`. But its definitely possible to transmit this information over the wire during initialization. Its too early for extensions but we can take this as a feature request.

Reviewer 3: Can we remove this until we need it or know how it should be defined? I understand this is the design documentation whereas there has been work done to implementation, but it should hold up to scrutiny we can always create issues to clean up the implementation.

Author: Sure.

Author: Opened up an issue: https://github.com/opensearch-project/opensearch-sdk/issues/65

>     javaVersion: '14'

Reviewer 3: Why is there a java version?  Is this a requirement for using a transport client?

Author: This is inherited from plugins, but you are right we dont need it.
@ryanbogan with everything removed do you know latest structure for `extensions.yml`?

    className: ExtensionsRunner
    customFolderName: opensearch-sdk
    hasNativeController: false	
```

### Communication

> TBD

Reviewer 3: Were you planning on filling these sections out in this PR or in a future PR?  If you were thinking about the future I might recommend creating the follow up issue and referencing it here.

Reviewer 2: Added it as a part of this PR

Author: @owaiskazi19 wanted to contribute :) which is why I left it as TBD.

### OpenSearch SDK
TBD

### Settings
Walking through a similar example as plugin above, after extension registration is done, extension makes an API call to register custom settings to OpenSearch.
`ExtensionsOrchestrator` recieves the requests, forwards it to `SettingsModule` to register a new setting and wala, the user is now able to toggle the setting via `_settings` Rest API.

## FAQ

- Will extensions replace plugins?  

>   We see value in plugins as they are high performant which works well for index/search workloads.

Reviewer 3: It would be simpler to say that plugins will continue to be supported and extensions are preferred as they will be easier to develop, deploy, and operate

Reviewer 4: +1 to @peternied comment

Author: I like it. I'll steal the comment :)

Extensions are designed to extend features via APIs which are exposed using extension points of OpenSearch.  

### Discovery
Extensions are discovered and configured via `extensions.yml` which is read by OpenSearch during the node bootstrap. `ExtensionsOrchestrator` reads through the config file at `~/extensions` and registers extensions within OpenSearch.
Here is an example extension configuration `extensions.yml`:

```
extensions:
  - name: opensearch-sdk
    uniqueId: opensearch-sdk-1
    hostName: 'sdk_host'
    hostAddress: '127.0.0.1'
    port: '4532'
    version: '1.0'
    description: Extension for the Opensearch SDK Repo
    opensearchVersion: '3.0.0'
    javaVersion: '14'
    className: ExtensionsRunner
    customFolderName: opensearch-sdk
    hasNativeController: false	
```

### Communication
As we are running extensions on the port defined in the `extensions.yml`, the communication between OpenSearch and Extensions happens using a ServerSocket which binds the port and the host address. OpenSearch will initialize the extensions during the bootstrap by making a request to all the extensions running on different ports and thus creating a medium for the future requests.

> ### OpenSearch SDK

Reviewer 3: This sections seems to be more about how to migrate plugins to extensions with this SDK, seems like it should be in another section or reshape to explain why there is an extension sdk.

Also I don't think this name `OpenSearch SDK` is useful - can we pick a better one like `opensearch-extension-sdk`? https://github.com/opensearch-project/opensearch-sdk/issues/13

Reviewer 3: Also, nix the word `currently` the document is in the present tense.

Author: @owaiskazi19 could you take care of it?

Reviewer 2: Should I update the section name to `Why OpenSearch SDK?`.
We can update the DESIGN.md with the right name of SDK repo once #13 is resolved.

Author: We could leave it for now. Since the issue is still not done. 
Whenever the renaming happens, lets make sure this doc is updated.

![](Docs/Extensions.png)

Extensions are independent processes which are built using `opensearch-sdk`. They communicate with OpenSearch via [transport](https://github.com/opensearch-project/OpenSearch/tree/main/modules/transport-netty4) protocol which today is used to communicate between OpenSearch nodes.  
Extensions are designed to extend features via transport APIs which are exposed using extension points of OpenSearch.

### Discovery
Extensions are discovered and configured via `extensions.yml`, same way we currently have `plugin-descriptor.properties` which is read by OpenSearch during the node bootstrap. `ExtensionsOrchestrator` reads through the config file at `~/extensions` and registers extensions within OpenSearch.
Here is an example extension configuration `extensions.yml`:

```
extensions:
  - name: opensearch-sdk // extension name

>     uniqueId: opensearch-sdk-1 // identifier for the extension

Reviewer 3: Shouldn't the name be the unique id?

Author: No its not. Unique Id helps when we'd like to run multiple instances of the same extension connected to 1 node of OpenSearch. For example, AD would like to have multiple extensions of itself running to train models etc but still be connected to the same node.

Author: Opened up an issue: https://github.com/opensearch-project/opensearch-sdk/issues/65

    hostName: 'sdk_host' // name of the host where extension is running

>     hostAddress: '127.0.0.1' // host to reach

Reviewer 3: Why both, shouldn't the name be resolved into the address?

Author: Ideally yes. As we are using the existing data structure `DiscoveryNode` [1]. 
We didnt really want to redefine everything Transport. 

[1] https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java#L126-L134

    port: '4532' // port to reach

>     version: '1.0' // extension version

Reviewer 3: Why isn't the version number provided by the extension during its handshake, it seems easy to get out of sync in this file that will likely not change after installation (unless this isn't the case?)

Author: Extensions are not hotswappable, this configuration file has to change anyway when the extension changes.
This is also one parameter which we use to validate/verify, are we talking to the right extension.

Reviewer 3: > Extensions are not hotswappable

How would the OpenSearch service enforce this? IMO this is a feature of the architecture

Author: Yeah its definitely one of the features we called out as part of https://github.com/opensearch-project/OpenSearch/issues/2283.
As we've discussed I'll open an issue to discuss on this item.

Author: Opened up an issue: https://github.com/opensearch-project/opensearch-sdk/issues/65
